### PR TITLE
Mc reweighting

### PR DESCRIPTION
**Pull request by maxnoe**
This adds a few functions handy for data mc comparisons:
* calculate equivalent observation time for proton mcs according to pdg measurement of cosmic ray sprectrum
* reweight gamma mcs to another spectrum and calculate obstime

### CODE REVIEW

File: fact/analysis/statistics.py [hunk 1]
```diff
@@ -1,6 +1,46 @@
...
+    Parameters
+    ----------
+    energy: number or array-like
+        energy points to evaluate
+    phi_0: float
+        Flux normalization
+    gamma: float
+        Spectral index
+    '''
+    return phi_0 * energy**(-gamma)
```
Comment (dneise) [inline fact/analysis/statistics.py]: this -(minus) is maximum surprise, no? 

I don't think putting the - into the docu is good either, the user should give -2.7 it she means -2.7

File: fact/analysis/statistics.py [hunk 2]
```diff
@@ -1,6 +1,46 @@
...
+    energy: number or array-like
+        energy points to evaluate
+    phi_0: float
+        Flux normalization
+    a: float
+        Parameter `a`  of the curved power law
+    b: float
+        Parameter `b` of the curved power law
+    '''
+    return phi_0 * energy ** (a + b * np.log10(energy))
```
Comment (dneise) [inline fact/analysis/statistics.py]: `-b` in docu `+b` in code .. maybe not good.

File: fact/analysis/statistics.py [hunk 3]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
         return significance[0]
 
     return significance
+
+
+def calc_weight_simple_to_curved(energy, gamma, a, b):
+    '''
+    Reweigt simulated events from a simulated power law with
```
Comment (dneise) [inline fact/analysis/statistics.py]: typo in reweight?

File: fact/analysis/statistics.py [hunk 4]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+    energy: float or array-like
+        Energy of the events
+    gamma: float
+        Spectral index of the simulated power law
+    a: float
+        Parameter `a` of the target curved power law
+    b: float
+        Parameter `b` of the target curved power law
+    '''
+    return energy ** (gamma + a + b * np.log10(energy))
```
Comment (dneise) [inline fact/analysis/statistics.py]: maybe crosscheck the signs ... given the sign questions from above

File: fact/analysis/statistics.py [hunk 5]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+        Spectral index of the simulated power law
+    target_gamma: float
+        Spectral index of the target power law
+    '''
+    return energy ** (target_gamma - simulated_gamma)
+
+
+def calc_gamma_obstime(n_events, gamma, phi_0, max_impact, e_min, e_max):
+    '''
+    Calculate the equivalent observation time for a gamma montecarlo set
```
Comment (dneise) [inline fact/analysis/statistics.py]: `gamma` refers sometimes to the particle type, sometimes to the spectral index ... proposal use `spectral_index` over `gamma`, and `flux_normalisation` over `phi_0` maybe.

File: fact/analysis/statistics.py [hunk 6]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+    Parameters
+    ----------
+    n_events: int
+        Number of simulated events
+    gamma: float
+        Spectral index of the simulated power law
+    phi_0: float
+        Flux normalization of the simulated power law
+    max_impact: float
+        Maximal simulated impact
```
Comment (dneise) [inline fact/analysis/statistics.py]: should units be given?

File: fact/analysis/statistics.py [hunk 7]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+    e_max: float
+        Maximal simulated energy
+    '''
+    if gamma >= -1:
+        raise ValueError('gamma must be < -1')
+
+    numerator = n_events * (1 - gamma)
+
+    t1 = phi_0 * max_impact**2 * np.pi
+    t2 = e_max**(1 - gamma) - e_min**(1 - gamma)
```
Comment (dneise) [inline fact/analysis/statistics.py]: maybe find names for `t1` and `t2`

File: fact/analysis/statistics.py [hunk 8]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+        Maximal simulated impact
+    e_min: float
+        Mimimal simulated energy
+    e_max: float
+        Maximal simulated energy
+    '''
+    if gamma >= -1:
+        raise ValueError('gamma must be < -1')
+
+    numerator = n_events * (1 - gamma)
```
Comment (dneise) [inline fact/analysis/statistics.py]: we don't understand this (1- gamma) factor here.

File: fact/analysis/statistics.py [hunk 9]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+    max_impact: float
+        Maximal simulated impact
+    viewcone: float
+        Viewcone in degrees
+    e_min: float
+        Mimimal simulated energy in GeV
+    e_max: float
+        Maximal simulated energy in GeV
+    '''
+    N0 = 1.8e4
```
Comment (dneise) [inline fact/analysis/statistics.py]: magic number?

Comment (dneise) [reply fact/analysis/statistics.py]: flux normalisation from PDG ? if so, please cite and give units maybe? Maybe also make it a default parameter

File: fact/analysis/statistics.py [hunk 10]
```diff
@@ -39,3 +79,100 @@ def li_ma_significance(n_on, n_off, alpha=0.2):
...
+
+    area = np.pi * max_impact**2
+    solid_angle = 2 * np.pi * (1 - np.cos(np.deg2rad(viewcone)))
+
+    numerator = n_events * (1 - gamma)
+    denominator = N0 * area * solid_angle * (e_max**(1 - gamma) - e_min**(1 - gamma))
```
Comment (dneise) [inline fact/analysis/statistics.py]: again, find some names for `numerator` and `denominator` and explain `(1 - gamma )` please.

**Review by dneise — DISMISSED** [review]
some open questions

File: fact/analysis/statistics.py (hunk 1 above)
Comment (maxnoe) [reply fact/analysis/statistics.py]: This is convention. In Astrophysics, gamma is positive in most papers and literature.

File: fact/analysis/statistics.py (hunk 2 above)
Comment (maxnoe) [reply fact/analysis/statistics.py]: Good catch

File: fact/analysis/statistics.py (hunk 3 above)
Comment (maxnoe) [reply fact/analysis/statistics.py]: Thx

File: fact/analysis/statistics.py (hunk 5 above)
Comment (maxnoe) [reply fact/analysis/statistics.py]: Good idea

**Comment by dneise** [comment]
This looks unit testable ... would you like to add some?

File: fact/analysis/statistics.py (hunk 8 above)
Comment (maxnoe) [reply fact/analysis/statistics.py]: Comes from integrating e**(-gamma)

**Comment by dneise** [comment]
8 lines docstring for a one-line math function seems to add more noise than it helps ... 
I am a friend of no docstring at all, if the function clearly does one thing, has a good name, and clear parameters.

File: fact/analysis/statistics.py (hunk 10 above)
Comment (maxnoe) [reply fact/analysis/statistics.py]: Names? There are no names for this. Its just the numerator and denominator of a large fraction.

**Comment by dneise** [comment]
We lack context here. Can you elaborate a bit for what one needs this functionality?

**Comment by maxnoe** [comment]
I renamed gamma, refactored a bit the proton obstime function. If you would be so kind to have a look again.

**Comment by maxnoe** [comment]
@dneise 

These features are needed for comparison of simulations and observations on a flux or event rate level, 
not just normed histograms.

To compare event rates of proton mcs with data, you need to know to how much observation times the simulated protons correspond. Hence the `calc_proton_obstime` function.

To be able to compare gamma-like data events with gamma monte carlos the same has to be done for gammas.
Additionally, since we observe different sources and do not simulate gammas with the respective source spectra we have to reweight the gammas to match the spectra of the source of interest.

**Review by jebuss — APPROVED** [review]
beautyful

**Pull request merged**